sources/pmwebapi.go

```diff
@@ -0,0 +1,164 @@
+package sources
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type GetToken struct {
+	Context int
+}
+
+type MetricHeader struct {
+	Metrics []PcpMetrics
+}
+
+type PcpMetrics struct {
+	Name        string
+	Pmid        int64
+	Indom       int64
+	TextOneline string `json:"text-oneline"`
+	TextHelp    string `json:"text-help"`
+	Sem         string
+	Units       string
+	Type        string
+}
+
+type TimestampHeader struct {
+	Timestamp
+	Values []GetValues
+}
+
+type Timestamp struct {
+	S  string
+	Us string
+}
+
+type GetValues struct {
+	Pmid      int64
+	Name      string
+	S         string
+	Us        string
+	Instances []InstanceList
+}
+
+type InstanceList struct {
+	Instance int64
+	Value    float64
+}
+
+type pcpPmwebapiMetric struct {
+	Labels      []string
+	Labelvalues []string
+	Name        string
+	TextHelp    string
+	Value       float64
+}
+
+func init() {
+	Factories["pmwebapi"] = newPcpSource
+}
+func gaugeMetric(labels []string, labelvalues []string, name string, textHelp string, value float64) prometheus.Metric {
+	return prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", name),
+			textHelp,
+			labels,
+			nil,
+		),
+		prometheus.GaugeValue,
+		value,
+		labelvalues...,
+	)
+}
+
+type pcpPmwebapiSource struct {
+	pcpPmwebapiMetrics []pcpPmwebapiMetric
+}
+
+func (p *pcpPmwebapiSource) Update(ch chan<- prometheus.Metric) error {
+	for _, metric := range p.pcpPmwebapiMetrics {
+		ch <- gaugeMetric(metric.Labels, metric.Labelvalues, metric.Name, metric.TextHelp, metric.Value)
+	}
+	return nil
+}
+func newPcpSource() (PcpSource, error) {
+
+	var p pcpPmwebapiSource
+
+	tokenResp, err := http.Get("http://localhost:44323/pmapi/context?hostname=localhost")
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
+		return nil, err
+	}
+	defer tokenResp.Body.Close()
+
+	tokenBody, _ := ioutil.ReadAll(tokenResp.Body)
+	tokenMap := make(map[string]float64)
+	err = json.Unmarshal([]byte(tokenBody), &tokenMap)
+
+	if err != nil {
+		log.Println(err)
+	}
+	contextNumber := strconv.FormatFloat(tokenMap["context"], 'f', -1, 64)
+
+	initialMetricsUrl := ("http://localhost:44323/pmapi/" + contextNumber + "/_metric")
+
+	getMetrics, err := http.Get(initialMetricsUrl)
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
+		return nil, err
+	}
+	defer getMetrics.Body.Close()
+
+	getMetricsBody, _ := ioutil.ReadAll(getMetrics.Body)
+
+	metricHeaderVar := &MetricHeader{}
```

Comment:
Based on offline discussions, you were close in getting this to work @kmada97. I quickly hacked together an ugly example here: https://play.golang.org/p/ELpRW2cTQ9

I will leave you to figure out which parts apply to your code and what you need to change.